app/src/org/commcare/dalvik/application/CommCareApplication.java

```diff
@@ -377,6 +424,96 @@ private int initializeAppResources(CommCareApp app) {
         }
     }
 
+    /**
+     * @return all installed ApplicationRecords, in alphabetical order
+     */
+    public ArrayList<ApplicationRecord> getInstalledAppRecords() {
+        ArrayList<ApplicationRecord> records = new ArrayList<>();
+        for (ApplicationRecord r : getGlobalStorage(ApplicationRecord.class)) {
+            records.add(r);
+        }
+        Collections.sort(records, new Comparator<ApplicationRecord>() {
+
+            @Override
+            public int compare(ApplicationRecord lhs, ApplicationRecord rhs) {
+                return lhs.getDisplayName().compareTo(rhs.getDisplayName());
+            }
+
+        });
+        return records;
+    }
+
+    /**
+     * @return all ApplicationRecords that are installed and NOT archived
+     */
+    public ArrayList<ApplicationRecord> getVisibleAppRecords() {
+        ArrayList<ApplicationRecord> visible = new ArrayList<>();
+        for (ApplicationRecord r : getInstalledAppRecords()) {
+            if (!r.isArchived()) {
+                visible.add(r);
+            }
+        }
+        return visible;
+    }
+
+    /**
+     * @return  all ApplicationRecords that are installed AND are not archived AND have MM verified
+     */
+    public ArrayList<ApplicationRecord> getReadyAppRecords() {
+        ArrayList<ApplicationRecord> ready = new ArrayList<>();
+        for (ApplicationRecord r : getVisibleAppRecords()) {
+            if (r.resourcesValidated()) {
+                ready.add(r);
+            }
+        }
+        return ready;
+    }
+
+    public boolean visibleAppsPresent() {
+        return getVisibleAppRecords().size() > 0;
+    }
+
+    /**
+     * @return the list of all installed apps as an array
+     */
+    public ApplicationRecord[] appRecordArray() {
+        ArrayList<ApplicationRecord> appList = CommCareApplication._().getInstalledAppRecords();
+        ApplicationRecord[] appArray = new ApplicationRecord[appList.size()];
+        int index = 0;
+        for (ApplicationRecord r : appList) {
+            appArray[index++] = r;
+        }
+        return appArray;
+    }
+
+    /**
+     * @param index an index into the global table of installed apps
+     * @return the ApplicationRecord that corresponds to the given index in the global list of
+     * installed apps
+     */
+    public ApplicationRecord getAppAtIndex(int index) {

```

Comment:
I'm not sure I understand the overall value of this method.

The installedAppRecords() call is not necessarily globally ordered, so this can cause problems if that changes. 

It seems like the save ways to do this are:
1: Retrieve the installedApprecords array within a context and iterate/select by index from there.
2: Retrieve the installedAppRecords array, and if the record needs to be passed somewhere through a marshalled interface, it is done by the app's unique ID.